nameko/dependencies.py

```diff
@@ -239,233 +267,76 @@ def injections(self):
-
-            shared_dependencies.setdefault(sharing_key, {})
-            instance = shared_dependencies[sharing_key].get(self.key)
-            if instance is None:
-                instance = self.dep_cls(*self.args, **self.kwargs)
-                shared_dependencies[sharing_key][self.key] = instance
-        else:
-            instance = self.dep_cls(*self.args, **self.kwargs)
-        instance.bind(name, container)
-
-        for name, attr in inspect.getmembers(instance):
-            if isinstance(attr, DependencyFactory):
-                prov = attr.create_and_bind_instance(name, container)
-                setattr(instance, name, prov)
-
-        self.instances.add(instance)
-        return instance
-
-
-def entrypoint(decorator_func):
-    """ Transform a callable into a decorator that can be used to declare
-    entrypoints.
-
-    The callable must return a DependencyFactory that creates the
-    EntrypointProvider instance.
+# metaclass to allow sharing key always?
+# or prefer explicit shared=None in all __init__s
+# or require/encourage **kwargs in __init__s
 
-    The returned ``wrapper`` function has a ``provider_cls`` attribute that
-    references the EntrypointProvider subclass returned by its factory. This
-    helps separate the ``@entrypoint`` decorated methods from the class
-    that implements the dependency - users only need to refer to the method
-    and nameko can determine the implementing class.
+class Entrypoint(Extension):
 
-    e.g::
-
-        @entrypoint
-        def http(bind_port=80):
-            return DependencyFactory(HttpEntrypoint, bind_port)
-
-        class Service(object):
-
-            @http
-            def foobar():
-                pass
-
-    """
-    def registering_decorator(wrapper, fn, args, kwargs):
-        """ Verify that ``decorator_func`` returns a DependencyFactory and
-        register ``fn`` as an entrypoint.
+    def __init__(self, *args, **kwargs):
+        # Entrypoints cannot be shared
+        kwargs.pop('shared', False)
+        super(Entrypoint, self).__init__(*args, **kwargs)
 
-        Saves a reference to the provider class of the factory onto
-        the ``wrapper`` function that wraps ``decorator_func``.
-        """
-        factory = decorator_func(*args, **kwargs)
-        if not isinstance(factory, DependencyFactory):
-            raise DependencyTypeError('Arguments to `entrypoint` must return '
-                                      'DependencyFactory instances')
-        register_entrypoint(fn, factory)
-        wrapper.provider_cls = factory.dep_cls
+    @classmethod
+    def entrypoint(cls, *args, **kwargs):
 
-        return fn
+        def registering_decorator(fn, args, kwargs):
+            instance = cls(*args, **kwargs)
+            register_entrypoint(fn, instance)
+            return fn
 
-    @wraps(decorator_func)
-    def wrapper(*args, **kwargs):
         if len(args) == 1 and isinstance(args[0], types.FunctionType):
```

Comment:
Wrapt only supports decorators with optional *kwargs*, not both args and kwargs. I guess the comparison we're doing below is considered icky - it means you can never write a decorator that genuinely accepts a function as its first arg e.g.

```python
def foo():
    pass

@bar(foo)
def baz():
    pass